torch/lib/THC/THCTensorIndex.cu

```diff
@@ -332,5 +332,63 @@ __global__ void indexSelectLargeIndex(TensorInfo<T, IndexType> dst,
   }
 }
 
+template <typename IndexType, unsigned int Dims>
+struct LinearIndexCalcData {
+  // sizes for Tensor dims (either from the Tensor, or the size of the adv indexer at that dim)
+  IndexType sizes[Dims];
+  IndexType strides[Dims];       // strides for Tensor
+  bool adv[Dims];                // which Tensors are advanced indexers
+  long *advIndexTensors[Dims];   // Adv Indexing Tensors
+};
+
+template <typename IndexType, unsigned int Dims>
+__device__ __forceinline__ long calculateOffset(
+  IndexType index,                  // index to calculate offset for
+  LinearIndexCalcData<IndexType, Dims> data
+)
+{
+  IndexType offset = 0;
+
+#pragma unroll
+  for (int dim = Dims - 1; dim >= 0; --dim) {
+    IndexType sizeAtDim, strideAtDim, indexAtDim, nextIndex;
+
+    strideAtDim = data.strides[dim];
+    sizeAtDim = data.sizes[dim];
+
+    if (data.adv[dim]) {
+      indexAtDim = data.advIndexTensors[dim][index % sizeAtDim];
+      if (dim > 0 && data.adv[dim - 1]) {
+        nextIndex = index;
+      } else {
+        nextIndex = index / sizeAtDim;
+      }
+    } else {
+      indexAtDim = index % sizeAtDim;
+      nextIndex = index / sizeAtDim;
+    }
+
+    offset += indexAtDim * strideAtDim;
+    index = nextIndex;
+  }
+
+  return offset;
+}
+
+template <typename IndexType, unsigned int Dims>
+__global__ void calculateLinearIndices(
+  long *output,               // output Tensor for indices
+  int elements,               // number of elements in output <-> indices to calculate
+  ptrdiff_t baseOffset,       // base offset into the Tensor
+  LinearIndexCalcData<IndexType, Dims> data
+)
+{
+  for (long i = blockIdx.x * blockDim.x + threadIdx.x;
+         i < elements;
+         i += blockDim.x * gridDim.x) {
+      output[i] = baseOffset + calculateOffset<IndexType, Dims>(i, data);
+   }
+}
+
```

Comment:
If `data.adv[dim]` was false for all dimensions then we wouldn't be doing advanced indexing.